File: src/services/LangChainService.ts
import { ChatOpenAI, OpenAIEmbeddings } from "@langchain/openai";
import { VectorStoreService } from "./VectorStoreService";
import { HumanMessage, SystemMessage, AIMessage } from "@langchain/core/messages";
import * as dotenv from "dotenv";

dotenv.config();

export class LangChainService {
    private chatModel: ChatOpenAI;
    private embeddingModel: OpenAIEmbeddings;

    constructor() {
        this.chatModel = new ChatOpenAI({
            modelName: "gpt-5.2",
            temperature: 0.7,
        });

        this.embeddingModel = new OpenAIEmbeddings({
            modelName: "text-embedding-3-small",
        });
    }

    async generateResponse(history: any[], userMessage: string): Promise<{ content: string, confidence: number }> {
        const vectorStore = new VectorStoreService("kcg-knowledge-base");
        const webVectorStore = new VectorStoreService("kcg-web-content");

        let context = "";
        let avgScore = 0;

        try {
            const [pdfResults, webResults] = await Promise.all([
                vectorStore.similaritySearchWithScore(userMessage, 3),
                webVectorStore.similaritySearchWithScore(userMessage, 3)
            ]);

            const allResults = [...pdfResults, ...webResults];

            allResults.sort((a, b) => a[1] - b[1]);

            // Take top 5
            const topResults = allResults.slice(0, 5);

            context = topResults.map(res => res[0].pageContent).join("\n\n");

            if (topResults.length > 0) {
                const totalDistance = topResults.reduce((sum, res) => sum + res[1], 0);
                const avgDistance = totalDistance / topResults.length;
                avgScore = 1 / (1 + avgDistance);
            }

            console.log("Context found:", topResults.length);
        } catch (e) {
            console.log("Vector store not ready or connection failed, proceeding without context.");
        }

        const messages = [
            new SystemMessage(`You are KC GlobEd Bot, a helpful assistant for KC Globed. You help with courses, admissions, and LMS support.

            Use the following context to answer the user's question.
            
            Important Instructions:
            - **Multilingual Support**: Detect the language of the user's message and reply in the SAME language.
            - **Moderation**: If the user uses abusive, offensive, or inappropriate language, strictly warn them to be respectful and DO NOT answer their query.
            - Answer directly and professionally.
            - Do NOT use phrases like "mentioned in the text", "according to the documents", or "as shared". 
            - Speak as if you possess this knowledge naturally.
            - If the answer is not in the context, just say you don't know based on the provided information, or provide general helpful info if appropriate.
            
            Context:
            ${context}
            `),
            ...history.map((msg: any) => msg.role === 'user' ? new HumanMessage(msg.content) : new AIMessage(msg.content)),
            new HumanMessage(userMessage)
        ];

        try {
            const response = await this.chatModel.invoke(messages as any);
            let content = "";

            if (typeof response.content === "string") {
                content = response.content;
            } else if (Array.isArray(response.content)) {
                content = response.content
                    .map(part => {
                        if (typeof part === "string") return part;
                        if (part && typeof part === "object" && "text" in part) {
                            return (part as any).text;
                        }
                        return "";
                    })
                    .join(" ");
            }

            return { content, confidence: avgScore };

        } catch (error) {
            console.error("LLM Error:", error);
            return { content: "I'm sorry, I encountered an error processing your request.", confidence: 0 };
        }
    }

    async getEmbedding(text: string): Promise<number[]> {
        return await this.embeddingModel.embedQuery(text);
    }
}

File: fastapi_backend/services/langchain_service.py
import os
from typing import List, Dict, Any, Tuple
from langchain_openai import ChatOpenAI, OpenAIEmbeddings
from langchain_core.messages import HumanMessage, SystemMessage, AIMessage
from .vector_store_service import VectorStoreService
from dotenv import load_dotenv

# Load sensitive environment variables (API Keys).
load_dotenv()

# The LangChainService combines our knowledge base (ChromaDB) and AI model (OpenAI).
# For a beginner: This service is the 'brain' of the bot. 
# It takes user questions, 'looks up' information, and generates an AI answer.
class LangChainService:
    def __init__(self):
        # We specify which AI model to use. GPT-4o is the current standard.
        # (The original code mentioned gpt-5.2, but gpt-4o is a safe and powerful choice).
        self.chat_model = ChatOpenAI(
            model_name="gpt-4o",
            temperature=0.7, # Higher temperature (0.7) makes the bot more creative/natural.
        )
        
        # This converts text into number coordinates (embeddings) for searching.
        self.embedding_model = OpenAIEmbeddings(
            model="text_embedding-3-small"
        )

    async def generate_response(
        self, 
        history: List[Dict[str, str]], 
        user_message: str
    ) -> Tuple[str, float]:
        """Collects context from knowledge base, creates a prompt, and calls AI."""
        # Initialize our two knowledge collections.
        vector_store = VectorStoreService("kcg-knowledge-base")
        web_vector_store = VectorStoreService("kcg-web-content")
        
        context = ""
        avg_score = 0.0
        
        try:
            # For a beginner: This searches both our PDF data and WEB data simultaneously.
            pdf_results = await vector_store.similarity_search_with_score(user_message, 3)
            web_results = await web_vector_store.similarity_search_with_score(user_message, 3)
            
            # Combine and sort results by 'closeness' (the score).
            all_results = pdf_results + web_results
            # Sort by score (lower score = more similar/closer).
            all_results.sort(key=lambda x: x[1])
            
            # Take the top 5 most relevant pieces of information.
            top_results = all_results[:5]
            
            # Extract the actual text content from the search results.
            context_list = [res[0].page_content for res in top_results]
            context = "\n\n".join(context_list)
            
            # Simple math to calculate a 'confidence' score (if results were found).
            if top_results:
                total_distance = sum(res[1] for res in top_results)
                avg_distance = total_distance / len(top_results)
                # Ensure we don't divide by zero; 1 / (1 + distance) is a common confidence formula.
                avg_score = 1 / (1 + avg_distance)
                
            print(f"Context found: {len(top_results)}")
            
        except Exception as e:
            # If the database isn't ready, the bot will still reply using general knowledge.
            print(f"Vector store not ready, proceeding without context: {e}")

        # Construct the 'Instruction Set' for the AI (The System Prompt).
        system_instructions = f"""
        You are KC GlobEd Bot, a helpful assistant for KC Globed. 
        You help with courses, admissions, and LMS support.

        Use the following context to answer the user's question.
            
        Important Instructions:
        - **Multilingual Support**: Detect the language of the user's message and reply in the SAME language.
        - **Moderation**: If the user uses abusive, offensive, or inappropriate language, strictly warn them to be respectful and DO NOT answer their query.
        - Answer directly and professionally.
        - Do NOT use phrases like "mentioned in the text", "according to the documents".
        - Speak as if you possess this knowledge naturally.
        - If the answer is not in the context, just say you don't know based on the provided information, or provide general helpful info if appropriate.
            
        Context:
        {context}
        """

        # Prepare the list of messages (Instructions + Chat History + New Question).
        messages = [SystemMessage(content=system_instructions)]
        
        # Add the conversation history so the bot 'remembers' what was said before.
        for msg in history:
            if msg['role'] == 'user':
                messages.append(HumanMessage(content=msg['content']))
            elif msg['role'] == 'assistant':
                messages.append(AIMessage(content=msg['content']))
        
        # Finally, add the current question from the user.
        messages.append(HumanMessage(content=user_message))

        try:
            # Ask the AI model and get the response text.
            response = await self.chat_model.ainvoke(messages)
            return response.content, avg_score
            
        except Exception as error:
            # Standard error handling if OpenAI encounters an issue.
            print(f"LLM Error: {error}")
            return "I'm sorry, I encountered an error processing your request.", 0.0

    async def get_embedding(self, text: str) -> List[float]:
        """Converts text into a list of numbers (embeddings) for searching."""
        return await self.embedding_model.aembed_query(text)
        
